vision.py

```diff
@@ -33,106 +33,123 @@ def run(self):
+    total_area = 0
+    for c in group:
+        area = contour_areas[c]
+        summed += contours[c].mean(axis=0)[0] * area
+        total_area += area
+    weighted_position =  summed / total_area
+    
+    return (
+        weighted_position[0] * 2.0 / frame_size[0] - 1.0,
+        weighted_position[1] * 2.0 / frame_size[1] - 1.0,
+    )
 
-    def filter_contours(self, contours: np.ndarray) -> np.ndarray:
-        """Unimplemented, returns identity
-        """
-        return contours
+def annotate_image(display: np.ndarray, contours: np.ndarray, group: List[int], pos: Tuple[int, int]) -> np.ndarray:
+    cv2.drawContours(
+        display,
+        contours,
+        -1,
+        (255, 0, 0),
+        thickness=2,
+    )
+    x = int((pos[0]+1)/2 * display.shape[1])
+    y = int((pos[1]+1)/2 * display.shape[0])
 
-    def group_contours(self, contours: np.ndarray) -> List[List[int]]:
-        """Returs a nested list of contour indices grouped by relative distance
-        """
-        # Build an adjacency list based on distance between contours relative to their area
-        connections = [[] for _ in contours]
-        for i, a in enumerate(contours):
-            a_com = a.mean(axis=0)[0]
-            a_area = cv2.contourArea(a)
-            for j, b in enumerate(contours[i + 1:]):
-                max_metric = max(a_area, cv2.contourArea(b)) * METRIC_SCALE_FACTOR
-                b_com = b.mean(axis=0)[0]
-                d = a_com - b_com
-                metric = d[0] ** 2 + d[1] ** 2
-                if (metric < max_metric):
-                    connections[i].append(j + i + 1)
-                    connections[j + i + 1].append(i)
-        # Breadth first search from each contour that wasn't yet assigned to a group to find its group
-        assigned = [False for _ in contours]
-        group = set()
-        groups = []
-        for i, c in enumerate(connections):
-            if assigned[i]: continue
-            group = {i}
-            assigned[i] = True
-            horizon = set(c)
-            while (True):
-                if len(horizon) == 0:
-                    # Group completed
-                    groups.append(list(group))
-                    break
-                for i in horizon: assigned[i] = True
-                group.update(horizon)
-                new_horizon = set()
-                for old in horizon:
-                    new_horizon.update(connections[old])
-                new_horizon.difference_update(group)
-                horizon = new_horizon
-        return groups
-
-    def rank_groups(self, contours: np.ndarray, groups: List[List[int]]) -> List[int]:
-        """Returns the group that is most likely to be the target
-        Could use metrics such as curvature, consistency of contours, consistency with last frame
-
-        Currently just returns the group with most elements
-        """
-        return max(groups, key=len)
+    for i, _ in enumerate(group[:-1]):
```

Comment:
Why `enumerate` but throw away the value?

It seems like a more appropriate thing to do is to `zip(group, group[1:])`.